### PR TITLE
Formalize lifetime of objects in herder

### LINKED ISSUE #509: Deal with when LRU pushes out things that SCP still needs

### PR DESCRIPTION
**Pull request by MonsieurNicolas**
# Description

Resolves #509 

This PR replaces #2327 

This PR adds wrapper objects to transaction sets and quorum sets, so that herder can keep an exact mapping between those objects and the ones referenced by SCP.

This is the first step in separating the caching strategy between SCP and herder/overlay: this should allow us to be a lot more aggressive in caching data that SCP doesn't strictly need in the future while keeping the strict minimum of data for SCP to continue to function.

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/herder/PendingEnvelopes.h [hunk 1]
```diff
@@ -65,14 +70,39 @@ class PendingEnvelopes
...
+    // extending the lifetime of the result
+    SCPQuorumSetPtr getKnownQSet(Hash const& hash, bool touch);
+
+    // NB: caller should use the return value to avoid duplicate tx sets
+    TxSetFramePtr putTxSet(Hash const& hash, uint64 slot, TxSetFramePtr txset);
+    // tries to find a txset in memory, setting touch also touches the LRU,
+    // extending the lifetime of the result
+    TxSetFramePtr getKnownTxSet(Hash const& hash, uint64 slot, bool touch);
+
+    void cleanKnownData();
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/herder/PendingEnvelopes.h]: I would strongly discourage use of `std::shared_ptr<TxSetFrame>(new TxSetFrame(txset))` because it can leak memory. If we _do_ want to avoid the control block issue, then we should allocate the pointer in a RAII fashion before transferring control to the `shared_ptr`.

File: src/scp/BallotProtocol.cpp [hunk 2]
```diff
@@ -1776,21 +1776,23 @@ BallotProtocol::setStateFromEnvelope(SCPEnvelope const& e)
     }
 }
 
-std::vector<SCPEnvelope>
-BallotProtocol::getCurrentState() const
+bool
+BallotProtocol::processCurrentState(
+    std::function<bool(SCPEnvelope const&)> const& f) const
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/BallotProtocol.cpp]: attempting to call a null `std::function` throws already, so we don't need to do this

File: src/scp/SCP.h [hunk 3]
```diff
@@ -105,8 +105,10 @@ class SCP
     // return highest slot index value
     uint64 getHighSlotIndex() const;
 
-    // returns all messages for the slot
-    std::vector<SCPEnvelope> getCurrentState(uint64 slotIndex);
+    // invokes f for all latest messages
+    // f returns false to stop processing, true otherwise
+    bool processCurrentState(uint64 slotIndex,
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/SCP.h]: SCP/slot yeah we don't use the return value, so I'll remove it (I think it was used in the other PR)

File: src/herder/HerderSCPDriver.cpp [hunk 4]
```diff
@@ -100,6 +100,53 @@ HerderSCPDriver::restoreSCPState(uint64_t index, StellarValue const& value)
...
+        if (!b)
+        {
+            if (throws)
+            {
+                throw std::runtime_error(
+                    "Invalid value in SCPHerderValueWrapper");
+            }
+            else
+            {
+                return;
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/HerderSCPDriver.cpp]: no: it's a best effort in the "no throw" version

File: src/herder/PendingEnvelopes.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/PendingEnvelopes.h]: I left this with the control block for now as the shared ptrs are of `TxSetFrame` and `SCPQuorumSet`, both of which are fairly small (they contain vectors that are destroyed right away so we're left with a few bytes)

**Comment by MonsieurNicolas** [comment]
@marta-lokhova addressed your feedback

File: src/herder/HerderSCPDriver.cpp [hunk 5]
```diff
@@ -737,7 +793,8 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
 
     // Ballot Protocol uses BASIC values
     comp.ext.v(STELLAR_VALUE_BASIC);
-    return xdr::xdr_to_opaque(comp);
+    auto res = wrapValue(comp);
```
Comment (jonjove) [inline src/herder/HerderSCPDriver.cpp]: I think this might race with line 786.

Comment (MonsieurNicolas) [reply src/herder/HerderSCPDriver.cpp]: yes, fixed

File: src/scp/BallotProtocol.h [hunk 6]
```diff
@@ -101,13 +101,13 @@ class BallotProtocol
     // c for EXTERNALIZE messages
     static SCPBallot getWorkingBallot(SCPStatement const& st);
 
-    SCPEnvelope*
+    SCPEnvelope const*
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/BallotProtocol.h]: probably not: there is no reason to keep a reference to this

File: src/scp/NominationProtocol.cpp [hunk 7]
```diff
@@ -626,7 +628,7 @@ NominationProtocol::getLatestMessage(NodeID const& id) const
     auto it = mLatestNominations.find(id);
     if (it != mLatestNominations.end())
     {
-        return &it->second;
+        return &it->second->env();
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/NominationProtocol.cpp]: same thing, no reason to have long lived references to the internals

File: src/scp/NominationProtocol.cpp [hunk 8]
```diff
@@ -26,6 +26,24 @@ NominationProtocol::NominationProtocol(Slot& slot)
 {
 }
 
+bool
+WrappedValuePtrComparator::operator()(ValueWrapperPtr const& l,
+                                      ValueWrapperPtr const& r) const
+{
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/NominationProtocol.cpp]: yeah we don't allow nullptr in there - I've added an assert

File: src/herder/HerderImpl.cpp [hunk 9]
```diff
@@ -515,25 +515,20 @@ HerderImpl::sendSCPStateToPeer(uint32 ledgerSeq, Peer::pointer peer)
         return;
     }
 
-    if (getSCP().getLowSlotIndex() > std::numeric_limits<uint32_t>::max() ||
-        getSCP().getHighSlotIndex() >= std::numeric_limits<uint32_t>::max())
+    auto itt = getSCP().ascSlots();
+    for (; itt.first != itt.second; itt.first++)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/HerderImpl.cpp]: I mentioned this to you earlier (not in PR): for this to work we'd have to use some sort of tri state callback as there are 3 possible return values:
* continue processing the slot
* stop processing the current slot, process the next slot
* stop processing everything

File: src/herder/PendingEnvelopes.h [hunk 10]
```diff
@@ -148,5 +149,10 @@ class PendingEnvelopes
 
     // updates internal state when an envelope was succesfuly processed
     void envelopeProcessed(SCPEnvelope const& env);
+
+    // queries state
+
+    bool isReady(SCPEnvelope const& e) const;
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/PendingEnvelopes.h]: those were left over from the other PR, I removed both of them

File: src/herder/HerderSCPDriver.cpp [hunk 11]
```diff
@@ -101,6 +101,18 @@ HerderSCPDriver::restoreSCPState(uint64_t index, StellarValue const& value)
 
 // envelope handling
 
+class SCPHerderEnvelopeWrapper : public SCPEnvelopeWrapper
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/HerderSCPDriver.cpp]: done

File: src/scp/SCPDriver.h [hunk 12]
```diff
@@ -16,6 +16,32 @@ namespace stellar
...
+{
+    SCPEnvelope mEnvelope;
+
+  public:
+    SCPEnvelopeWrapper()
+    {
+    }
+
+    virtual ~SCPEnvelopeWrapper();
+    virtual void setEnvelope(SCPEnvelope const& e);
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/SCPDriver.h]: I was trying to avoid throwing from constructors, but the base class doesn't throw so I'm removing those set methods

File: src/herder/HerderSCPDriver.cpp [hunk 13]
```diff
@@ -100,6 +100,53 @@ HerderSCPDriver::restoreSCPState(uint64_t index, StellarValue const& value)
...
+    }
+
+    void
+    setValue(Value const& value) override
+    {
+        setValue(value, false);
+    }
+
+    void
+    setValue(Value const& value, bool throws)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/HerderSCPDriver.cpp]: this is correct, I through I was going to be more ambitious on this PR by using this in pending envelope but as I didn't do it, I'll keep the strict versions only

File: src/herder/test/HerderTests.cpp [hunk 14]
```diff
@@ -1514,7 +1514,8 @@ TEST_CASE("In quorum filtering", "[quorum][herder][acceptance]")
             HerderImpl& herder = *static_cast<HerderImpl*>(&c->getHerder());
 
             auto const& lcl = c->getLedgerManager().getLastClosedLedgerHeader();
-            herder.getSCP().processCurrentState(lcl.header.ledgerSeq, proc);
+            herder.getSCP().processCurrentState(lcl.header.ledgerSeq, proc,
+                                                true);
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/test/HerderTests.cpp]: the test is making an assertion on all messages from all peers (previous version was not correct)

File: src/overlay/test/FloodTests.cpp [hunk 15]
```diff
@@ -277,13 +277,15 @@ TEST_CASE("Flooding", "[flood][overlay][acceptance]")
...
+                lcl.header.ledgerSeq + 1,
+                [&](SCPEnvelope const& e) {
                     if (keysMap.find(e.statement.nodeID) != keysMap.end())
                     {
                         okCount++;
                     }
                     return true;
-                });
+                },
+                true);
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/overlay/test/FloodTests.cpp]: the test is making an assertion on all messages from all peers (previous version was not correct)

File: src/herder/HerderImpl.cpp [hunk 16]
```diff
@@ -841,16 +843,19 @@ HerderImpl::resolveNodeID(std::string const& s, PublicKey& retKey)
...
+                    std::string curK = KeyUtils::toStrKey(e.statement.nodeID);
+                    if (curK.compare(0, arg.size(), arg) == 0)
+                    {
+                        retKey = e.statement.nodeID;
+                        r = true;
+                        return false;
+                    }
+                    return true;
+                },
+                true);
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/HerderImpl.cpp]: we're looking for any message from any peer (including the local node), so there should not be any filtering

**Comment by MonsieurNicolas** [comment]
@jonjove addressed your feedback. Thanks!

File: src/scp/SCPDriver.h [hunk 17]
```diff
@@ -16,6 +16,32 @@ namespace stellar
...
+
+    SCPEnvelope const&
+    env() const
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/SCPDriver.h]: done

File: src/scp/SCPDriver.h [hunk 18]
```diff
@@ -16,6 +16,28 @@ namespace stellar
 {
 typedef std::shared_ptr<SCPQuorumSet> SCPQuorumSetPtr;
 
+class SCPEnvelopeWrapper
+{
+    SCPEnvelope const mEnvelope;
+
+  public:
+    SCPEnvelopeWrapper(SCPEnvelope const& e);
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/SCPDriver.h]: yeah I agree

File: src/scp/SCPDriver.h [hunk 19]
```diff
@@ -16,6 +16,28 @@ namespace stellar
...
+    virtual ~SCPEnvelopeWrapper();
+
+    SCPEnvelope const&
+    GetEnvelope() const
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/SCPDriver.h]: search/replace operator error :)

File: src/herder/HerderImpl.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/HerderImpl.cpp]: oh I didn't realize that this is what you were going for, I thought you wanted to expose slots.
Change is fine with me

File: src/herder/HerderSCPDriver.h [hunk 20]
```diff
@@ -144,6 +146,9 @@ class HerderSCPDriver : public SCPDriver
     bool checkCloseTime(uint64_t slotIndex, uint64_t lastCloseTime,
                         StellarValue const& b) const;
 
+    // wraps a *valid* StellarValue (throws if it can't find txSet/qSet)
+    ValueWrapperPtr wrapValue(StellarValue const& sv);
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/herder/HerderSCPDriver.h]: :facepalm:

File: src/scp/NominationProtocol.cpp [hunk 21]
```diff
@@ -298,30 +298,30 @@ NominationProtocol::getNodePriority(NodeID const& nodeID,
...
         {
-            valueToNominate = value;
+            valueToNominate = mSlot.getSCPDriver().wrapValue(value);
         }
         else
         {
             valueToNominate = extractValidValue(value);
         }
-        if (!valueToNominate.empty())
+        if (valueToNominate)
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/NominationProtocol.cpp]: yes, comment for `extractValidValue` (SCPDriver.h) already says
```
// returning nullptr means no valid value could be extracted
```

File: src/scp/SCPDriver.h [hunk 22]
```diff
@@ -40,6 +40,24 @@ class WrappedValuePtrComparator
 
 typedef std::set<ValueWrapperPtr, WrappedValuePtrComparator> ValueWrapperPtrSet;
 
+struct SCPBallotWrapper
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/scp/SCPDriver.h]: didn't realize I left it there, I moved it into `BallotProtocol` as it's the only place we use it and cached the `SCPBallot` per your suggestion

File: src/scp/NominationProtocol.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (jonjove) [reply src/scp/NominationProtocol.cpp]: I see, you meant that as an if-and-only-if. Sounds good, thanks.

**Comment by jonjove** [comment]
r+ e191fc4c

File: src/herder/PendingEnvelopes.cpp [hunk 23]
```diff
@@ -563,18 +563,13 @@ PendingEnvelopes::slotClosed(uint64 slotIndex)
...
-
-    return TxSetFramePtr();
+    return getKnownTxSet(hash, 0, true);
 }
 
 SCPQuorumSetPtr
 PendingEnvelopes::getQSet(Hash const& hash)
 {
-    auto qset = getKnownQSet(hash, true);
+    auto qset = getKnownQSet(hash, false);
```
Comment (jonjove) [inline src/herder/PendingEnvelopes.cpp]: Why did you change this to false? This doesn't match the behavior of `getTxSet` and it also doesn't match the old behavior of `getQSet` prior to this PR.

**Comment by jonjove** [comment]
r+ c61a7bb
